<a href="https://www.kaggle.com/code/bhagwati/spaceship-titanic-for-begineers?scriptVersionId=145297014" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

## Introduction

Welcome to the basic level analysis of SpaceShip Titanic Dataset and submission of our prediction for competition. We will cover this beginner level analysis into different section which outlined below :

#### 1: Importing and Reading Dataset
#### 2: Description of Dataset
#### 3: Missing values Analysis 
#### 4: Encoding & Transforming Features 
#### 5: Builidng Models
#### 6: Output Submission

#### Importing and Reading Dataset

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/spaceship-titanic/sample_submission.csv
/kaggle/input/spaceship-titanic/train.csv
/kaggle/input/spaceship-titanic/test.csv


<br><br>
Reading the dataset using pandas and storing train and test datasets into different variables for further analysis. <br> df_train and df_test are two pandas dataframe for
training data and test data respectively.

In [2]:
df_train = pd.read_csv('../input/spaceship-titanic/train.csv')
df_test = pd.read_csv('../input/spaceship-titanic/test.csv')

In [3]:
# First Five rows from dataframe default value is 5.
# we can get any number of rows like head(10) to get first 10 rows.
df_train.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

#### Description of Dataset

In [4]:
# train and test data size
print("Shape of train data : ",df_train.shape)
print("Shape of test data : ",df_test.shape)

#Just to print dotted separation line 
print("-"*70)

# List of all the columns names
print(df_train.columns.values)
print(df_test.columns.values)

Shape of train data :  (8693, 14)
Shape of test data :  (4277, 13)
----------------------------------------------------------------------
['PassengerId' 'HomePlanet' 'CryoSleep' 'Cabin' 'Destination' 'Age' 'VIP'
 'RoomService' 'FoodCourt' 'ShoppingMall' 'Spa' 'VRDeck' 'Name'
 'Transported']
['PassengerId' 'HomePlanet' 'CryoSleep' 'Cabin' 'Destination' 'Age' 'VIP'
 'RoomService' 'FoodCourt' 'ShoppingMall' 'Spa' 'VRDeck' 'Name']


##### Observation :
1. Total entries are 8693 and no. of cols is 14(including target one).
2. There are 7 objects(string or mixed),6 float type and 1 is boolean.
3. All cols have null values except PassengerId and Transported.

In [5]:
# Information of dataset and their columns
# like Datatypes of columns,Null values,total entries etc. 

df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8693 entries, 0 to 8692
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   PassengerId   8693 non-null   object 
 1   HomePlanet    8492 non-null   object 
 2   CryoSleep     8476 non-null   object 
 3   Cabin         8494 non-null   object 
 4   Destination   8511 non-null   object 
 5   Age           8514 non-null   float64
 6   VIP           8490 non-null   object 
 7   RoomService   8512 non-null   float64
 8   FoodCourt     8510 non-null   float64
 9   ShoppingMall  8485 non-null   float64
 10  Spa           8510 non-null   float64
 11  VRDeck        8505 non-null   float64
 12  Name          8493 non-null   object 
 13  Transported   8693 non-null   bool   
dtypes: bool(1), float64(6), object(7)
memory usage: 891.5+ KB


In [6]:
#statistical description of dataset like mean,meadian,percentile etc for all the columns
df_train.describe()

Age   RoomService     FoodCourt  ShoppingMall           Spa  \
count  8514.000000   8512.000000   8510.000000   8485.000000   8510.000000   
mean     28.827930    224.687617    458.077203    173.729169    311.138778   
std      14.489021    666.717663   1611.489240    604.696458   1136.705535   
min       0.000000      0.000000      0.000000      0.000000      0.000000   
25%      19.000000      0.000000      0.000000      0.000000      0.000000   
50%      27.000000      0.000000      0.000000      0.000000      0.000000   
75%      38.000000     47.000000     76.000000     27.000000     59.000000   
max      79.000000  14327.000000  29813.000000  23492.000000  22408.000000   

             VRDeck  
count   8505.000000  
mean     304.854791  
std     1145.717189  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%       46.000000  
max    24133.000000

In [7]:
# to check if dataset is balance or imbalance
df_train['Transported'].value_counts()

Transported
True     4378
False    4315
Name: count, dtype: int64

*This is balanced dataset as we saw above that True values are 4378 and False are 4315 which is nearly equal.*

#### Missing Value Analysis

Every Dataset we get,is not always clean and good to use.So first we need to take a 
look at our dataset for certain checks.<br>As we saw in info method above,dataset have missing or null values in all the columns excpet PassengerId & Tranported. 
we need to check count of null values column wise. 

1. CryoSleep col has most missing values 217 and 2.49% of total entries.
2. PassengerId and Transported has no missing values at all.

In [8]:
# count of missing values column wise in descending order
print(df_train.isnull().sum().sort_values(ascending=False))

# Just to print dotted separation line
print("-"*30)

# percentage of the missing values in the column Cryosleep(which has most missing values)
print("missing value % :",((df_train['CryoSleep'].isnull().sum())/len(df_train))*100)

CryoSleep       217
ShoppingMall    208
VIP             203
HomePlanet      201
Name            200
Cabin           199
VRDeck          188
FoodCourt       183
Spa             183
Destination     182
RoomService     181
Age             179
PassengerId       0
Transported       0
dtype: int64
------------------------------
missing value % : 2.4962613597147127


##### Handling missing values
We need to handle missing values either by removing columns or filling missing entries.Here,max missing values is 2.49% so we need not to remove any column ,Just to fill those entries with proper filling strategy.
1. Numerical features cols should be filled with mean or median of that particular column.
2. Categorical will be filled with mode of that column.
<br>

#### Transforming Features <br>
Age, RoomService, FoodCourt, ShoppingMall, Spa, VRDeck are the features,which has numerical values.so we need to fill with either mean or median.Here I am filling those with median of respective columns. Applying on train and test data both.

In [9]:
l = ['Age','RoomService','FoodCourt','ShoppingMall','Spa','VRDeck']
for i in l:
    df_train[i].fillna(df_train[i].median(),inplace=True)
    df_test[i].fillna(df_test[i].median(),inplace=True)

HomePlanet, CryoSleep, Destination, VIP are the features which have not numerical values so we can't use mean or median to fill as these can be calculated for numerical values only.So here we are using mode(most frequent value of respective col) to fill missing values.

In [10]:
#filling categorical features with mode of respective columns
m = ['HomePlanet','CryoSleep','Destination','VIP']
for i in m:
    df_train[i].fillna(df_train[i].mode()[0],inplace=True)
    df_test[i].fillna(df_test[i].mode()[0],inplace=True)

In [11]:
#Removing transported column from training dataset and storing it in to y.
y = df_train['Transported']
df_train.drop('Transported',axis=1,inplace=True)

We have Cabin and Name features in our dataset but These two features are not very relevant for training purpose,so we are removing these two features from our training and test data.

In [12]:
# Drop Cabin and Name column from train data
df_train.drop(columns=['Cabin','Name'],axis=1,inplace=True)

# Drop Cabin and Name column from test data
df_test.drop(columns=['Cabin','Name'],axis=1,inplace=True)

We have CryoSleep and VIP features which has values in string as True and False.But system will not understand this string type.so we need to change True to 1 and False to 0. We can do it by changing data type as integer.

In [13]:
# Change true -> 1 and False -> 0 
df_train['CryoSleep']=df_train['CryoSleep'].astype('int')
df_train['VIP']=df_train['VIP'].astype('int')

df_test['CryoSleep']=df_test['CryoSleep'].astype('int')
df_test['VIP']=df_test['VIP'].astype('int')

#### Encoding 
Encoding is a technique to convert categorical data to numerical so that machine can understand and train it.Here we have HomePlanet and Destination which have categorical data,showing below :-

In [14]:
# count of all the distinct values in HomePlanet(a categorical feature)
print("train",df_train['HomePlanet'].value_counts())
print("-"*30)
print("test",df_test['HomePlanet'].value_counts())

train HomePlanet
Earth     4803
Europa    2131
Mars      1759
Name: count, dtype: int64
------------------------------
test HomePlanet
Earth     2350
Europa    1002
Mars       925
Name: count, dtype: int64


Here we are using one of the encoding technique that is LabelEncoding.It will covert all the categorical values to numerical ones like -
for Homeplanet - There are three categories - Earth, Europe and Mars.so it will convert these three into 0,1 and 2.

In [15]:
# LabelEncoding for train and test data for Homeplanet and Destination Features
from sklearn.preprocessing import LabelEncoder
Le = LabelEncoder()
df_train['HomePlanet']=Le.fit_transform(df_train['HomePlanet'])
df_train['Destination']=Le.fit_transform(df_train['Destination'])

df_test['Destination']=Le.fit_transform(df_test['Destination'])
df_test['HomePlanet']=Le.fit_transform(df_test['HomePlanet'])

#### Building Models <br>
First we need to split our df_train dataset into train and test using train_test_split method.<br>
Test_size has set 0.3 it means df_train will be splitted into two parts.In which 70% is training data and 30% is test data.

In [16]:
# splitting the train dataset into train and test dataset using train_test_split method
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(df_train,y,random_state=20,test_size=0.3)

In [17]:
#RandomForest classifier for training dataset
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
rm_f = RandomForestClassifier(max_depth=7,random_state=0)
rm_f.fit(X_train,y_train)
a = rm_f.predict(X_test)
print(accuracy_score(a,y_test))

0.7994631901840491


In [18]:
# Logistic regression classifier for training
from sklearn.linear_model import LogisticRegression
log_reg = LogisticRegression(random_state=0)
log_reg.fit(X_train,y_train)
b = log_reg.predict(X_test)
print(accuracy_score(b,y_test))

0.7879601226993865


In [19]:
# DecisionTree classifier for training 
from sklearn.tree import DecisionTreeClassifier
dt_clf = DecisionTreeClassifier(random_state=0)
dt_clf.fit(X_train,y_train)
c = dt_clf.predict(X_test)
print(accuracy_score(c,y_test))

0.713957055214724


#### Output Submission <br>
As we have seen above that best accuracy score is for RandomForest Classifier and that is around 80%.so making submission on behlaf on randomforest model.<br>
Assiging results of randomforest model to y_pred and passengerid to id. y_pred and id are the variables which we are going to place in submission file.

In [20]:
y_pred = rm_f.predict(df_test)
id = df_test.PassengerId

In [21]:
# Reading the format of sample submission file,same format we need to follow for our submission
sub = pd.read_csv('/kaggle/input/spaceship-titanic/sample_submission.csv')
sub.head()

PassengerId  Transported
0     0013_01        False
1     0018_01        False
2     0019_01        False
3     0021_01        False
4     0023_01        False

In [22]:
#Making Dataframe of id and predicted values
submit = pd.DataFrame({"PassengerId":id,"Transported":y_pred})

In [23]:
# Submitting our results to output.csv and making index false for ignoring default index in output
submit.to_csv('output.csv',index=False)

**Thanks for going through my work and I happy to get some reviews,suggestion and questions to improve my understanding.**